### DISCUSSION TITLE
Displaying inline Python documentation in CoCalc for Jupyter notebooks

### DISCUSSION THREAD

**Discussion by miohtama** [sagemathinc/cocalc#6714] [Q&A] [answered]
What would be the best option to enable displaying inline documentation for functions a notebook user is using?

- Usually, editors and IDEs do this on mouse hover
- I remember Jupyter Notebooks have had some alternative approaches for this in the past

For my use case, the documentation is Sphinx style docstring from Python function (same as for Pandas, others.)

I would like this to appear when I put a cursor on a function.

How it works in PyCharm when editing notebooks:

<img width="1484" alt="image" src="https://github.com/sagemathinc/cocalc/assets/49922/a5e91441-af1e-4f6c-9e30-1bba5d0218fa">

Also, if there is a way to auto-import for functions or offer auto-import, that would be great, as these functions are usually related.

Autoimport hint in PyCharm:

<img width="984" alt="image" src="https://github.com/sagemathinc/cocalc/assets/49922/4f94ecf9-261f-406c-a406-9c8e5d57fcfb">

**Comment by williamstein** [comment]
> What would be the best option to enable displaying inline documentation for functions a notebook user is using?

This is not yet implemented in CoCalc.   When it gets implemented it will probably be in conjunction with LSP support.  See https://github.com/sagemathinc/cocalc/issues/5217

You can display docstring in the standard Jupyter notebook way, by typing "shift+tab" with the cursor right after the opening parenthesis.  The dostring then appears in a separate panel, which you can maximize, split, etc.  But it's not inline.

<img width="777" alt="image" src="https://github.com/sagemathinc/cocalc/assets/1276278/ab5b0735-381b-4968-beea-da00e79086f1">

> Also, if there is a way to auto-import for functions or offer auto-import, that would be great, as these functions are usually related.

Interesting.  I've never heard of this before! How exactly does it work?  Does it add code in the cell, then run the cell?  There are many potential design choices for such a feature, so maybe you could describe exactly what pycharm does and whether or not you like it all of it.

**Comment by miohtama** [comment]
PyCharm, and I assume their Datalore product for data research as well, does the auto import by following

- It is the same for normal Python code and Jupyter notebooks
- They use their own LSP to know all your Python code in your project
- When you type an unidenfied expression you get an option to auto-correct it - and PyCharm will add the correct import
- PyCharm also keeps added imports in a correct Pythonic order and does not add it to the last
- You get an option to import in the current cell, or more options  

<img width="873" alt="image" src="https://github.com/sagemathinc/cocalc/assets/49922/6215069d-ea22-4f31-a427-5f0cd0e2aa35">

<img width="1059" alt="image" src="https://github.com/sagemathinc/cocalc/assets/49922/36dd0d50-4ab4-4159-b1a5-c71540ac96d0">


PyCharm's auto import seems to be 100x more efficient and reliable e.g. one we get in Visual Studio Code through Pylance, and it s a major productivity boost when writing any Python.

**Comment by miohtama** [comment]
> This is not yet implemented in CoCalc. When it gets implemented it will probably be in conjunction with LSP support. See https://github.com/sagemathinc/cocalc/issues/5217

Just to clarify on this - Is the current `jupyter-lsp` incompatible with CoCalc?

**Reply by williamstein** [reply, marked as answer]
>  Is the current jupyter-lsp incompatible with CoCalc?

Cocalc-jupyter is a completely different stack top-to-bottom than the python/jupyterlab ecosystem.
